src/sparsebase/io/mtx_writer.cc

```diff
@@ -0,0 +1,320 @@
+#include "sparsebase/format/csr.h"
+
+namespace sparsebase::io {
+
+template <typename IDType, typename NNZType, typename ValueType>
+MTXWriter<IDType, NNZType, ValueType>::MTXWriter(
+    std::string filename, std::string object, std::string format, std::string field, std::string symmetry)
+    : filename_(filename),
+      object_(object),
+      format_(format),
+      field_(field),
+      symmetry_(symmetry) {}
+
+template <typename IDType, typename NNZType, typename ValueType>
+void MTXWriter<IDType, NNZType, ValueType>::WriteCOO(
+    format::COO<IDType, NNZType, ValueType> *coo) const {
+      
+      //illegal parameter checks
+      if (object_ != "matrix" && object_ != "vector") 
+      {
+        throw utils::ReaderException("Illegal value for the 'object' option in matrix market header");
+      }
+      else if (object_ == "vector")
+      {
+        throw utils::ReaderException("Matrix market writer does not currently support writing vectors.");
+      }
+      if (format_ != "array" && format_ != "coordinate") 
+      {
+        throw utils::ReaderException("Illegal value for the 'format' option in matrix market header");
+      }
+      if (field_ != "real" && field_ != "double" && field_ != "complex" && field_ != "integer" && field_ != "pattern") 
+      {
+        throw utils::ReaderException("Illegal value for the 'field' option in matrix market header");
+      }
+      if (symmetry_ != "general" && symmetry_ != "symmetric" && symmetry_ != "skew-symmetric" && symmetry_ != "hermitian") 
+      {
+        throw utils::ReaderException("Illegal value for the 'symmetry' option in matrix market header");
+      }
+      if (format_ == "array" && field_ == "pattern") 
+      {
+        throw utils::ReaderException("Matrix market files with array format cannot have the field ""'pattern' ");
+      }
+      if (format_ == "array" && symmetry_ != "general") 
+      {
+        throw utils::ReaderException("Matrix market files with array format cannot have the property 'symmetry' ");
+      }
+      if (symmetry_ == "hermitian") 
+      {
+        throw utils::ReaderException("Matrix market writer does not currently support hermitian symmetry.");
+      }
+
+      std::ofstream mtxFile;
+      mtxFile.open(filename_);
+      
+      //header
+      mtxFile << "%%MatrixMarket " << object_ << " " << format_ << " " << field_ << " " << symmetry_ << "\n";
+      
+
+      //TODO add warning when given field and actual data format is not same (integer vs float etc)
+
+      if constexpr (std::is_same_v<ValueType, void>)
+      {
+        throw utils::WriterException("Cannot write an MTX with void ValueType");
+      }
+      else
+      {
+        auto dimensions = coo->get_dimensions();
+        IDType* row = coo->get_row();
+        IDType* col = coo->get_col();
+        ValueType* val = coo->get_vals();
+
+        if (val != NULL && field_ == "pattern")
+        {
+          std::cerr << "Warning: Pattern selected but values given.\n";
```

Comment:
You should use the logger here instead of std::cerr.

Reply:
fixed.